### PR TITLE
Improvement/openweathermap

### PR DESCRIPTION
**Pull request by kiddigital**
This PR fixes the 'broken' Open Weather Map integration.

As Open Weather Map also released a new and improved version of the API, this PR also switches to this new API.

Multiple new Domoticz sensor have been added and improved. Now also Wind Gust, Chill temperature, UV Index, Weather description and 'feel' temperature are available as well. Now 9 sensors for the current weather situation.

Daily forecast added, showing the forecast for Today and 7 more days. Each day has 10 sensors (so 80 in total). Each day has for example min- max temperature and probability of Precipitation (changes of something coming down from the clouds in any shape or form like rain, snow, drizzle, hail. etc.). Also some textual sensor to build a dashboard, etc.

Also hourly forecast data processed if available for 48 hours. With 8 sensors per hour a total of 384 sensor are generated!

Multiple choices can be made to (not) use certain parts like current weather and/or forecast data and the polling interval can be provided (600 seconds by default).

Main reason to get the Open Weather Map integration back up and running is the disappearance of many services for actual free weather data.
- DarkSky has been acquired by Apple and will stop (API at least, Android already since 1-8-2020) by the end of 2021
- Wunderground (as part of The Weather Company, owned by IBM since 2016) also stopped there API's
- AccuWeather is focussed on US mainly.
- PWSWeather has a lot of free data from Personal Weather Stations, but is also US owned (AerisWeather)

Future ideas/improvements:
- You can also feed data from your own weather station, so the idea is to have different local Domoticz sensor become the input sensors for your local OWM weather station 
- Support for a lot of great weather Icons that could be utilized
- Generate local weather maps (not sure if animated) to display in Domoticz
- and more ...?? 

Comments/suggestions/bugs/etc. feel free to put in the comments...

### CODE REVIEW

File: hardware/DomoticzHardware.cpp [hunk 1]
```diff
@@ -792,7 +792,8 @@ void CDomoticzHardwareBase::SendPercentageSensor(const int NodeID, const uint8_t
 	_tGeneralDevice gDevice;
 	gDevice.subtype = sTypePercentage;
 	gDevice.id = ChildID;
-	gDevice.intval1 = NodeID;
+	gDevice.battery_level = BatteryLevel;
+	gDevice.intval1 = (NodeID << 8) | ChildID;
```
Comment (gizmocuz) [inline hardware/DomoticzHardware.cpp]: This will cause a lot of new Node ID's for various hardware. Please undo this change

Comment (gizmocuz) [reply hardware/DomoticzHardware.cpp]: Battery Percentage is passed along with sDecodeRXMessage, not needed here

File: hardware/DomoticzHardware.cpp [hunk 2]
```diff
@@ -822,6 +823,7 @@ void CDomoticzHardwareBase::SendVisibilitySensor(const int NodeID, const int Chi
 	_tGeneralDevice gDevice;
 	gDevice.subtype = sTypeVisibility;
 	gDevice.id = ChildID;
+	gDevice.battery_level = BatteryLevel;
```
Comment (gizmocuz) [inline hardware/DomoticzHardware.cpp]: Battery Percentage is passed along with sDecodeRXMessage, not needed here

File: main/mainworker.cpp [hunk 3]
```diff
@@ -10095,6 +10095,7 @@ void MainWorker::decode_General(const CDomoticzHardwareBase* pHardware, const tR
 		(subType == sTypePercentage) ||
 		(subType == sTypeWaterflow) ||
 		(subType == sTypePressure) ||
+		(subType == sTypeVisibility) ||
```
Comment (gizmocuz) [inline main/mainworker.cpp]: This change will cause other hardware to have new ID's, please undo

File: www/app/hardware/Hardware.html [hunk 4]
```diff
@@ -979,6 +979,39 @@ <h2><span data-i18n="Hardware">Hardware</span>:</h2>
...
+			<tr>
+				<td align="right" style="width:110px"><label for="adddayforecast"><span data-i18n="Daily forecast">Daily forecast</span>:</label></td>
+				<td><input type="checkbox" id="adddayforecast" />&nbsp;<label for="adddayforecast">&nbsp;Forecast for today and coming 7 days  (if available).</label></td>
+			</tr>
+			<tr>
+				<td align="right" style="width:110px"><label for="addhourforecast"><span data-i18n="Hourly forecast">Hourly forecast</span>:</label></td>
+				<td><input type="checkbox" id="addhourforecast" />&nbsp;<label for="addhourforecast">&nbsp;Forecast for each hour for 48 hours (if available).</label></td>
+			</tr>
+			<tr>
+				<td align="right" style="width:110px"><label for="pollinterval"><span data-i18n="Poll Interval"></span>:</label></td>
```
Comment (gizmocuz) [inline www/app/hardware/Hardware.html:1010]: I don't think this should be a user defined option here. It might cause hammering and could result in a ban.
It also does not make any sense to poll faster then once every 5 minutes

File: www/app/hardware/Hardware.html [hunk 5]
```diff
@@ -979,6 +979,39 @@ <h2><span data-i18n="Hardware">Hardware</span>:</h2>
...
+					<span data-i18n="(0 = Application default)"></span>
+				</td>
+			</tr>
+			<tr>
+				<td align="right" style="width:110px"><label for="addcurrent"><span data-i18n="Current Weather">Current Weather</span>:</label></td>
+				<td><input type="checkbox" id="addcurrent" checked="checked" />&nbsp;<label for="addcurrent">&nbsp;</label></td>
+			</tr>
+			<tr>
+				<td align="right" style="width:110px"><label for="adddayforecast"><span data-i18n="Daily forecast">Daily forecast</span>:</label></td>
+				<td><input type="checkbox" id="adddayforecast" />&nbsp;<label for="adddayforecast">&nbsp;Forecast for today and coming 7 days  (if available).</label></td>
```
Comment (gizmocuz) [inline www/app/hardware/Hardware.html]: Noticed extra space before (if avail...)

File: www/app/hardware/Hardware.js [hunk 6]
```diff
@@ -2055,7 +2089,37 @@ define(['app'], function (app) {
 					}
 				});
 			}
-			else if (text.indexOf("Buienradar") >= 0) {
```
Comment (gizmocuz) [inline www/app/hardware/Hardware.js:2058]: Just to be sure, Buienradar settings are still working ?

**Review by gizmocuz — COMMENTED** [review]
Thank you for this PR and work !
Please review my review

File: hardware/DomoticzHardware.cpp (hunk 1 above)
Comment (kiddigital) [reply hardware/DomoticzHardware.cpp]: > 
> 
> This will cause a lot of new Node ID's for various hardware. Please undo this change

True, will revert.

The behavior of the generated NodeID for the Percentage Sensor is different from a lot of other sensors because the ChildID is not part of the generated ID. Could be considered a 'bug'?

File: main/mainworker.cpp (hunk 3 above)
Comment (kiddigital) [reply main/mainworker.cpp]: For the Visibility sensor the NodeID is also not using the ChildID, sort of same as for Percentage. Maybe for consistency somehow make these sensors behave the same?

Will revert for this PR.

File: www/app/hardware/Hardware.html (hunk 4 above)
Comment (kiddigital) [reply www/app/hardware/Hardware.html:1010]: Good point, Did considered it also and there is some sanity checking the values in the constructor (around line 80) with a minimum of 10 seconds and max of 86400 (a day). OWM has a limit of 60 per second! and 1mln per day :) But agree that some sensible values are needed.

I will set the minimum to 60 seconds and default remains 600 seconds.

File: www/app/hardware/Hardware.js (hunk 6 above)
Comment (kiddigital) [reply www/app/hardware/Hardware.js:2058]: Everything still works as far as I could see and test :)

**Comment by kiddigital** [comment]
@gizmocuz , review comments have been addressed. All should be fine now (I hope :) )

File: hardware/DomoticzHardware.cpp (hunk 1 above)
Comment (gizmocuz) [reply hardware/DomoticzHardware.cpp]: It is not a 'bug' but it is how it was designed to work. There are more sensors. I guess at that time i thought 255 percentage sensors should be enough per hardware ;)

**Comment by gizmocuz** [comment]
Thanks for this PR and your hard work !!!

**Comment by xztraz** [comment]
Awesome work!

I read someone is missing the poll intervall setting since he got a warning on too many calls via api.
https://www.domoticz.com/forum/viewtopic.php?f=12&t=35153

is every sensor having its own api call now? could the calls be bunched together?

**Comment by kiddigital** [comment]
@xztraz , all sensors are updated using just 1 single API call.

The ability to change the poll frequency has been left out after discussing this with @gizmocuz 

Maybe we could think about a settings 'lower update frequency' that updates the sensors at half the standard frequency? That means an OWM module only performs half of the amount of API calls it does now.

**Comment by gizmocuz** [comment]
Or just create another account ?

**Comment by xztraz** [comment]
i've not got that error myself(one installation). maybe the guy in the forum is using same api on many installations. if only one call per update there should not be any problem.
